v2/pkg/engine/plan/interface_selection_rewriter.go

```diff
@@ -0,0 +1,359 @@
+package plan
+
+import "github.com/wundergraph/graphql-go-tools/v2/pkg/ast"
+
+type interfaceSelectionRewriter struct {
+	operation  *ast.Document
+	definition *ast.Document
+
+	fieldRef        int
+	enclosingNode   ast.Node
+	dsConfiguration DataSourceConfiguration
+}
+
+func newInterfaceSelectionRewriter(operation *ast.Document, definition *ast.Document) *interfaceSelectionRewriter {
+	return &interfaceSelectionRewriter{
+		operation:  operation,
+		definition: definition,
+	}
+}
+
+func (r *interfaceSelectionRewriter) isFieldReturnsInterface(fieldRef int, enclosingNode ast.Node) (interfaceDefRef int, isInterface bool) {
+	var (
+		fieldDefRef int
+		hasField    bool
+	)
+
+	interfaceDefRef = ast.InvalidRef
+
+	switch enclosingNode.Kind {
+	case ast.NodeKindObjectTypeDefinition:
```

Comment:
Is this even necessary? You dismiss object types in line 51.

Reply:
field could be on an interface or an object
```
type EnclosingObject {
  field: PossibleInterfaceType
}

type EnclosingInterface {
  field: PossibleInterfaceType
}
```

on line 51 I'm checking if a field type e.g. `PossibleInterfaceType` is an interface